### ISSUE TITLE
debug protocol: Set next line

### ISSUE THREAD

**Issue by dzzie** [svaarala/duktape#243]
Would this be possible with the pc2line data? 

I have always found this to be a fairly useful feature in visual studio C, C# & vb6.
allows you to reset execution to a specific line (within the same scope) so you can
retest results without having to end your current debug session.

an edit and continue feature would kick butt too, but thats probably going to be pretty 
complex/bloated. feature. (i imagine you can replace an entire function through by overridding
it with the debugger eval will have to test that)

**Comment by svaarala** [comment]
That should be relatively straightforward to implement; one could either provide a primitive to set the next line or the next PC to execute. There are some exceptions though - e.g. if you jump into (or out of) a try-catch or an iteration statement, Duktape state would be incorrect because the book-keeping happens by executing certain opcodes explicitly. Similarly, if one were to jump to an arbitrary PC (in the middle of a statement), unsafe things might happen.

In any case this would be quite useful so let's keep this open for the next debugger feature set.

**Comment by fatcerberus** [comment]
I would guess that, in order to implement this completely safely, Duktape would have to maintain some sort of source map along with the bytecode, so that setting the PC to the middle of a statement would be impossible (doing so would round it to the nearest statement).  Of course that would greatly increase footprint, so might not be feasible.

**Comment by svaarala** [comment]
Yeah, since statements don't always correspond 1:1 to lines, using the pc2line map for this would not always work. One would essentially need a statement list and the initial PC for each statement. After that the issue would be try-catch-blocks and the like.

**Comment by dzzie** [comment]
or you could limit it to only where a pc2line was valid. If that had 95% coverage users could just accept that it didnt support certain constructs and modify thier code accordingly if they needed it in that specific circumstance. (like multiple statements on one line etc)

**Comment by fatcerberus** [comment]
@dzzie In theory pc2line is always valid.  An error could be thrown at any point during execution; for error augmentation to work, the engine has to have some way to map an arbitrary PC value to a line number.

**Comment by dzzie** [comment]
doh makes sense, somehow i was thinking in reverse on that one
